src/lib/transaction_snark_scan_state/transaction_snark_scan_state.ml

```diff
@@ -636,3 +636,38 @@ let fill_work_and_enqueue_transactions t transactions work =
         else Or_error.error_string "Unexpected ledger proof emitted" )
   in
   (result_opt, updated_scan_state)
+
+let required_state_hashes _t =
+  (* TODO: when merging into #4244
+  List.map (Parallel_scan.pending_data t) ~f:(fun (t : Transaction_with_witness.t) -> ...*)
+  State_hash.Set.empty
+
+let check_required_protocol_states t ~protocol_states =
+  let open Or_error.Let_syntax in
+  let required_state_hashes = required_state_hashes t in
+  let check_length states =
+    let required = State_hash.Set.length required_state_hashes in
+    let received = List.length states in
+    if required = received then Or_error.return ()
+    else
+      Or_error.errorf
+        !"Required %d protocol states but received %d"
+        required received
+  in
+  (*Don't check further if the lengths dont match*)
+  let%bind () = check_length protocol_states in
+  let received_state_map =
+    List.fold protocol_states ~init:Coda_base.State_hash.Map.empty
+      ~f:(fun m ps ->
+        State_hash.Map.set m ~key:(Coda_state.Protocol_state.hash ps) ~data:ps
+    )
+  in
+  let protocol_states_assoc =
+    List.filter_map (State_hash.Set.to_list required_state_hashes)
+      ~f:(fun hash ->
+        let open Option.Let_syntax in
+        let%map state = State_hash.Map.find received_state_map hash in
+        (hash, state) )
+  in
+  let%map () = check_length protocol_states_assoc in
```

Comment:
Nit: this would be more efficient as

```suggestion
  let received_state_keys = State_hash.Map.key_set received_state_map in
  assert (State_hash.Set.equal required_state_hashes received_state_map) ;
  let protocol_states_assoc = Map.to_list received_state_map in
```

Reply:
Efficient because it is not looking up in the map?
Also, `key_set` is not present in the version of core_kernel (0.12) we use

Reply:
Sorry, must have checked the wrong docs.

Current version is `n*log2(n)+2n` because of the `Map.find`, adjusted version is `3*n`, and failure is earlier in the adjusted version. If we don't have `key_set` then it's not worth the effort though :shrug: Apologies for the noise.